pkl-core/src/main/java/org/pkl/core/runtime/ResourceManager.java

```diff
@@ -165,8 +167,38 @@ public Optional<Object> read(URI resourceUri, Node readNode) {
 
           throw new VmExceptionBuilder()
               .evalError("unsupportedResourceType", reader.getClass().getName(), res.getClass())
-              .withLocation(readNode)
+              .withOptionalLocation(readNode)
               .build();
         });
   }
+
+  /**
+   * Used by ResourceReaders.ProjectPackageResource to resolve resources from projects that may not
+   * be on the local filesystem
+   */
+  public List<PathElement> listElements(URI baseUri) throws IOException, SecurityManagerException {
+    var reader = resourceReaders.get(baseUri.getScheme());
+    if (reader == null) {
+      throw new VmExceptionBuilder()
+          .evalError("noResourceReaderRegistered", baseUri.getScheme())
+          .build();
+    }
+
+    return reader.listElements(securityManager, baseUri);
+  }
+
+  /**
+   * Used by ResourceReaders.ProjectPackageResource to resolve resources from projects that may not
+   * be on the local filesystem
+   */
+  public boolean hasElement(URI elementUri) throws IOException, SecurityManagerException {
+    var reader = resourceReaders.get(elementUri.getScheme());
+    if (reader == null) {
+      throw new VmExceptionBuilder()
+          .evalError("noResourceReaderRegistered", elementUri.getScheme())
+          .build();
+    }
+
+    return reader.hasElement(securityManager, elementUri);
+  }
```

Comment:
I am not a huge fan of needing to pipe these through `ResourceManager` like this. I'm more than open to alternate suggestions on how to handle this